crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -2,163 +2,231 @@ use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
 use anyhow::anyhow;
-use group::{secp256k1, GroupElement};
+use group::secp256k1::group_element::Value;
+use group::{secp256k1, GroupElement, PartyID};
+use im::hashmap;
+use itertools::Itertools;
 use maurer::knowledge_of_discrete_log::PublicParameters;
 use maurer::Proof;
 use pera_types::base_types::{AuthorityName, ObjectID};
+use pera_types::error::{PeraError, PeraResult};
 use pera_types::event::Event;
 use pera_types::messages_consensus::ConsensusTransaction;
 use proof::mpc::{AdvanceResult, Party};
 use rand_core::OsRng;
+use rayon::prelude::*;
+use schemars::_private::NoSerialize;
+use std::cmp::PartialEq;
 use std::collections::{HashMap, VecDeque};
-use std::io;
+use std::future::Future;
 use std::io::Write;
 use std::marker::PhantomData;
 use std::sync::{Arc, Weak};
 use std::time::Duration;
+use std::{io, mem};
 use tokio::sync::{mpsc, Mutex, RwLock};
 use tokio::time::sleep;
 use tracing::{debug, error, info};
 
-#[derive(Debug)]
-pub enum MPCInput {
-    InitEvent(CreatedProofMPCEvent),
-    Message,
+#[derive(Clone)]
+struct ProofMPCMessage {
+    message: Vec<u8>,
+    authority: AuthorityName,
+}
+pub type ProofMessage = (Proof<1, Lang, PhantomData<()>>, Vec<Value>);
+
+fn authority_name_to_party_id(
+    authority_name: AuthorityName,
+    epoch_store: &AuthorityPerEpochStore,
+) -> PeraResult<PartyID> {
+    Ok(epoch_store
+        .committee()
+        .authority_index(&authority_name)
+        // This should never happen, as the validator only accepts messages from committee members
+        .ok_or_else(|| {
+            PeraError::InvalidCommittee(
+                "Received a proof MPC message from a validator that is not in the committee"
+                    .to_string(),
+            )
+        })? as PartyID)
 }
 
 struct MPCInstance {
     status: MPCSessionStatus,
-    /// The channel to send message to this instance
-    input_receiver: Option<mpsc::Sender<MPCInput>>,
-    pending_messages: Vec<MPCInput>,
+    pending_messages: HashMap<PartyID, ProofMessage>,
     consensus_adapter: Arc<dyn SubmitToConsensus>,
     epoch_store: Weak<AuthorityPerEpochStore>,
     /// The threshold number of parties required to participate in each round of the Proof MPC protocol
     mpc_threshold_number_of_parties: usize,
+    session_id: ObjectID,
+    party: Option<ProofParty>,
 }
 
 type ProofPublicParameters =
     maurer::language::PublicParameters<{ maurer::SOUND_PROOFS_REPETITIONS }, Lang>;
 
 impl MPCInstance {
-    fn set_active(&mut self, public_parameters: ProofPublicParameters) {
-        self.status = MPCSessionStatus::Active;
-        // TODO (#256): Replace hard coded 100 with the number of validators times 10
-        let (messages_handler_sender, messages_handler_receiver) = mpsc::channel(100);
-        self.input_receiver = Some(messages_handler_sender);
-        self.spawn_mpc_messages_handler(public_parameters, messages_handler_receiver);
+    async fn new(
```

Comment:
This function does a lot of things (such as setting the status to `Active` immediately and generating the first
message); can we have some documentation for the function describing this?